### DISCUSSION TITLE
correct way of setting cookies & headers in server actions

### DISCUSSION THREAD

**Discussion by kombos** [vercel/next.js#64096] [App Router]
I'm building a NextJS 14.1.3 app, where i'm using server action for form submission which validates an OTP at the backend, and responds back with cookies set. In the backend, the NextJS Server actually forwards the OTP to another Resource Server (built using playframework and scala) using the postValidateOtp() function (shown below), which internally uses fetch function:

```
export async function otpAction(username, otp) {
  let playResponse;
  const cookieStore = cookies();

  try {
    playResponse = await postValidateOtp(username, otp);
    // take 'Set-Cookie' header value
    const playResponseCookies = playResponse.headers.get("set-cookie");
    // set set-cookie header string to response cookie
    if (playResponseCookies) {
      let cookieObject = parseHeaderToCookie(playResponseCookies);
      cookieStore.set(cookieObject);
    }

    if (!playResponse.ok) {
      let playResponseBody = await playResponse.json();
      if (
        ["VALIDATE_OTP_FAILED", "INVALID_OTP"].some((str) =>
          playResponseBody?.message?.includes?.(str)
        )
      )
        return { ok: false, status: 401 };
      return { ok: false, status: playResponse?.status };
    }
  } catch (error) {
    return { ok: false, status: 500 };
  }

  // dont return anything since can't return non-serializable object like fetch response
  return { ok: true, status: playResponse?.status || 500 };
}
```
this is my postValidateOtp() function: 

```
export const postValidateOtp = async (usernameString, otpString) => {
  const reqBodyJson = {
    mobile: usernameString?.toString?.(),
    otp: otpString?.toString?.(),
  };

  // fetch the POST route from resource server
  const playFrameworkResponse = await fetch(validateOTPUrl, {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
    },
    body: JSON.stringify(reqBodyJson),
  });

  if (!playFrameworkResponse.ok) {
    console.error(
      "Error during postValidateOtp: ",
      playFrameworkResponse?.status
    );
  }

  return playFrameworkResponse;
};
```

The postValidateOtp() is a HTTP POST request to the resource server, which returns back with 'Set-Cookie' headers. I need to get those cookies and set it to the response back from the server action. As you can see, currently i'm manually parsing the cookie to an object and setting it using cookies().set(). Is there a better way to directly set all headers and cookies received in response from resource server (playframework) to the response returned from nextjs server to client?

**Comment by icyJoseph** [comment]
I think the way you have it is the way to go, as far as cookies handling is concerned that is.
